core/matrix/ell.hpp

```diff
@@ -0,0 +1,271 @@
+
+    void move_to(Dense<ValueType> *other) override;
+
+    void read_from_mtx(const std::string &filename) override;
+
+    /**
+     * Returns the values of the matrix.
+     *
+     * @return the values of the matrix.
+     */
+    value_type *get_values() noexcept { return values_.get_data(); }
+
+    /**
+     * @copydoc Ell::get_values()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type *get_const_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * Returns the column indexes of the matrix.
+     *
+     * @return the column indexes of the matrix.
+     */
+    index_type *get_col_idxs() noexcept { return col_idxs_.get_data(); }
+
+    /**
+     * @copydoc Ell::get_col_idxs()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type *get_const_col_idxs() const noexcept
+    {
+        return col_idxs_.get_const_data();
+    }
+
+    /**
+     * Returns the maximum number of non-zeros per row.
+     *
+     * @return the maximum number of non-zeros per row.
+     */
+    size_type get_max_nonzeros_per_row() const noexcept { return max_nonzeros_per_row_; }
+
+    /**
+     * Returns the padding of the matrix.
+     *
+     * @return the padding of the matrix.
+     */
+    size_type get_padding() const noexcept { return padding_; }
+
+    /**
+     * Returns the `idx`-th non-zero element of the `row`-th row .
+     *
+     * @param row  the row of the requested element
+     * @param col  the column of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU matrix from
+     *        the CPU results in a runtime error)
+     */
+    value_type &val_at(size_type row, size_type col) noexcept
+    {
+        return values_.get_data()[linearize_index(row, col)];
+    }
+
+    /**
+     * @copydoc Ell::val_at(size_type, size_type)
+     */
+    value_type val_at(size_type row, size_type col) const noexcept
+    {
+        return values_.get_const_data()[linearize_index(row, col)];
+    }
+
+    /**
+     * Returns the `idx`-th column index of the `row`-th row .
+     *
+     * @param row  the row of the requested element
+     * @param col  the column of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU matrix from
+     *        the CPU results in a runtime error)
+     */
+    index_type &col_at(size_type row, size_type col) noexcept
+    {
+        return get_col_idxs()[linearize_index(row, col)];
```

Comment:
nit: a part of our code style is to call methods within a class through the `this` pointer, this way it is clear that the output of the method also depends on the state of the current object.
In this case it would be:
```c++
return this->get_col_idxs()[this->linearize_index(row, col)];
```

I do realize this is not honoured in `gko::matrix:Dense`, so there's no way you could have figured it out. I will fix it in `Dense`.

Reply:
fixed